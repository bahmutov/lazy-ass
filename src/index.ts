function isArrayLike(a) {
  return a && typeof a.length === 'number'
}

function toStringArray(arr) {
  return (
    'array with ' +
    arr.length +
    ' items.\n[' +
    arr.map(toString).join(',') +
    ']\n'
  )
}

function isPrimitive(arg) {
  return (
    typeof arg === 'string' ||
    typeof arg === 'number' ||
    typeof arg === 'boolean'
  )
}

function isError(e) {
  return e instanceof Error
}

/*
  custom JSON.stringify replacer to make sure we do not
  hide properties that have value "undefined"
  var o = {
    foo: 42,
    bar: undefined
  }

  standard JSON.stringify returns
    '{"foo": 42}'
  this replacer returns
    '{"foo": 42, "bar": null}'
*/
function replacer(key, value) {
  if (value === undefined) {
    return null
  }
  return value
}

function toString(arg, k) {
  if (arg === null) {
    return 'null'
  }
  if (arg === undefined) {
    return 'undefined'
  }
  if (isPrimitive(arg)) {
    return JSON.stringify(arg)
  }
  if (arg instanceof Error) {
    return arg.name + ' ' + arg.message
  }

  if (Array.isArray(arg)) {
    return toStringArray(arg)
  }
  if (isArrayLike(arg)) {
    return toStringArray(Array.prototype.slice.call(arg, 0))
  }
  var argString
  try {
    argString = JSON.stringify(arg, replacer, 2)
  } catch (err) {
    argString =
      '{ cannot stringify arg ' + k + ', it has type "' + typeof arg + '"'
    if (typeof arg === 'object') {
      argString += ' with keys ' + Object.keys(arg).join(', ') + ' }'
    } else {
      argString += ' }'
    }
  }
  return argString
}

function endsWithNewLine(s) {
  return /\n$/.test(s)
}

function formMessage(args) {
  var msg = args.reduce(function(total, arg, k) {
    if (k && !endsWithNewLine(total)) {
      total += ' '
    }
    if (typeof arg === 'string') {
      return total + arg
    }
    if (typeof arg === 'function') {
      var fnResult
      try {
        fnResult = arg()
      } catch (err) {
        // ignore the error
        fnResult = '[function ' + arg.name + ' threw error!]'
      }
      return total + fnResult
    }
    var argString = toString(
      arg,
      k
    )
    return total + argString
  }, '')
  return msg
}

function lazyAssLogic(condition: any, ...args: any[]) {
  if (isError(condition)) {
    return condition
  }

  var fn = typeof condition === 'function' ? condition : null

  if (fn) {
    condition = fn()
  }
  if (!condition) {
    // var args = [].slice.call(arguments, 1);
    if (fn) {
      args.unshift(fn.toString())
    }
    return new Error(formMessage(args))
  }
}

interface LazyAss {
  (condition: any, ...args: any[]): void
  async: (condition: any, ...args: any[]) => void
}

export const lazyAss: LazyAss = <LazyAss>(
  function lazyAss(condition: any, ...args: any[]) {
    var err = lazyAssLogic(condition, ...args)
    if (err) {
      throw err
    }
  }
)

export const lazyAssync = function lazyAssync(condition: any, ...args: any[]) {
  var err = lazyAssLogic(condition, ...args)
  if (err) {
    setTimeout(function() {
      throw err
    }, 0)
  }
}

lazyAss.async = lazyAssync

export default lazyAss
