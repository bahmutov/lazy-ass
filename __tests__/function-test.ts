import { lazyAss } from '../src';

describe('function as condition', function() {
  it('evaluates the function', function() {
    var called;
    function condition() {
      called = true;
      return true;
    }

    lazyAss(condition);
    expect(called).toBe(true)
  });

  it('no result is failure', function() {
    function noreturn() {}

    expect(function() {
      lazyAss(noreturn);
    }).toThrowError()
  });

  it('adds condition function source to message', function() {
    function myCondition() {}
    expect(function() {
      lazyAss(myCondition);
    }).toThrowError(/myCondition/);
  });

  it('allows anonymous functions', function() {
    var called;
    lazyAss(function() {
      return true;
    });
    lazyAss(function() {
      return true;
    }, 'everything is ok');
    lazyAss(function() {
      called = true;
      return true;
    }, 'everything is ok');
    expect(called).toBe(true);

    expect(function() {
      lazyAss(function() {});
    }).toThrowError();
  });

  it('has access via closure', function() {
    var foo = 2,
      bar = 3;
    expect(function() {
      lazyAss(function() {
        return foo + bar === 6;
      }, 'addition');
    }).toThrowErrorMatchingSnapshot()
  });

  it('example', function() {
    var foo = 2,
      bar = 2;
    function isValidPair() {
      return foo + bar === 4;
    }
    lazyAss(isValidPair, 'foo', foo, 'bar', bar);
  });
});
