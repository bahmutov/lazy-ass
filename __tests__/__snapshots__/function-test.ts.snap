// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function as condition has access via closure 1`] = `
"function () {
                return foo + bar === 6;
            } addition"
`;
